usr/src/uts/common/fs/zfs/zfs_vnops.c

```diff
@@ -4424,6 +4424,46 @@ zfs_inactive(vnode_t *vp, cred_t *cr, caller_context_t *ct)
 	rw_exit(&zfsvfs->z_teardown_inactive_lock);
 }
 
+static void
+zfs_inactive_task(void *task_arg)
+{
+	vnode_t *vp = (vnode_t *)task_arg;
+	ASSERT(vp);
+	zfs_inactive_impl(vp, CRED(), NULL);
+}
+
+/*
+ * This value will be multiplied by zfs_dirty_data_max to determine
+ * the threshold past which we will call zfs_inactive_impl() async.
+ *
+ * Selecting the multiplier is a balance between how long we're willing to wait
+ * for delete/free to complete (get shell back, have a NFS thread captive, etc)
+ * and reducing the number of active requests in the backing taskq.
+ *
+ * 4 GiB (zfs_dirty_data_max default) * 16 (multiplier default) = 64 GiB
+ * meaning by default we will call zfs_inactive_impl async for vnodes > 64 GiB
+ */
+uint16_t zfs_inactive_async_multiplier = 16;
+
+void
+zfs_inactive(vnode_t *vp, cred_t *cr, caller_context_t *ct)
+{
+	znode_t	*zp = VTOZ(vp);
+
+	if (zp->z_size > zfs_inactive_async_multiplier * zfs_dirty_data_max) {
+		if (taskq_dispatch(dsl_pool_vnrele_taskq(
+		    dmu_objset_pool(zp->z_zfsvfs->z_os)), zfs_inactive_task,
+		    vp, TQ_SLEEP) != NULL)
```

Comment:
Seems like we might want to use TQ_NOSLEEP so that if there are a ton of deletions, and the queue gets full, then we will effectively recruit more threads to work on deletion (because the calling thread will work on deletion, rather than waiting for the taskq threads to make progress).